azure-pipelines.yml

```diff
@@ -17,50 +17,72 @@ variables:
 
-steps:
-- task: UsePythonVersion@0
-  inputs:
-    versionSpec: '3.9'
-  displayName: Use Python 3.9
+stages:
 
-- script: python -m pip install --upgrade pip
-  displayName: 'Install tools'
+- stage: unitTests
+  displayName: Run Unit Tests
+  jobs:
+  - job: unitTests
+    displayName: Run Unit Tests
 
-- script: pip install -r requirements.txt
-  displayName: 'Install requirements'
+    steps:
+      - task: UsePythonVersion@0
+        inputs:
+          versionSpec: '3.9'
+        displayName: 'Use Python 3.9'
 
-- script: |
-    pip install pytest
-    pip install pytest-cov
-    pytest tests --junitxml=junit/test-results.xml --cov=. --cov-report=xml --cov-report=html
-  displayName: 'Test with pytest'
+      - script: |
+          export PIP_NO_CACHE_DIR=off \
+            PIP_DISABLE_PIP_VERSION_CHECK=on \
+            PIP_DEFAULT_TIMEOUT=100 \
+            POETRY_VERSION=1.1.4 \
+            POETRY_VIRTUALENVS_IN_PROJECT=true \
+            POETRY_NO_INTERACTION=1
+        displayName: 'Export environment variables'
+      
+      - script: |
+          sudo apt-get install -y \
+            curl \
+            build-essential \
+            python3-distutils \
+            python3-apt \
+            libpython3.9-dev
+          curl -sSL https://raw.githubusercontent.com/sdispater/poetry/master/get-poetry.py | python -
+          source $HOME/.poetry/env
+          poetry install
+        displayName: 'Install dependencies'
 
-- task: Docker@2
-  displayName: 'Docker build'
-  condition: and(succeeded(), eq(variables.isMaster, true))
-  inputs:
-    repository: $(imageName)
-    command: 'build'
-    Dockerfile: 'Dockerfile'
-    tags: |
-      $(Build.SourceBranchName)-$(fullSha)
+      - script: |
+          source $HOME/.poetry/env
+          poetry run pytest --cov=data_service/
+        displayName: 'Test with pytest'
 
-# Authenticate Docker to GCR using predefined service connection
-- task: Docker@2
-  displayName: 'Login to GCR'
-  condition: and(succeeded(), eq(variables.isMaster, true))
-  inputs:
-    command: login
-    containerRegistry: gcrServiceConnection-Microdatano
+- stage: build
+  displayName: Build Docker Image and push to GCR
+  dependsOn: unitTests
+  # Run only if previous stage succeeded. Don't run for tagging events
+  condition: succeeded()
+  jobs:
+    - job: buildAndPush
+      displayName: Build and Push to GCR
 
-# Push Docker Image to GCR
-- task: Docker@2
-  displayName: 'Push image to GCR'
-  condition: and(succeeded(), eq(variables.isMaster, true))
-  inputs:
-    ContainerRegistry: |
-      gcrServiceConnection-Microdatano
-    repository: $(repoName)
-    command: 'push'
-    tags: |
-      $(Build.SourceBranchName)-$(fullSha)
+      steps:
+        # Authenticate Docker to GCR using predefined service connection
+        - task: Docker@2
+          displayName: 'Login to GCR'
+          condition: and(succeeded(), eq(variables.isMaster, true))
```

Comment:
This condition and the one on line 81 can be removed and only line 64 can to have it.

Reply:
Yes, I think we only need to build and push Docker images from master branch.